### ISSUE TITLE
Holoviews / Datashader integration

### ISSUE THREAD

**Issue by Huite** [Deltares/xugrid#204]
Holoviews and Datashader supports TriMesh/trimesh (resp.):
https://datashader.org/user_guide/Trimesh.html

This is obviously useful, especially for larger grids.

Althought not every 2D grid is triangular, it can be easily triangulated (which is done for a number of plotting methods, since matplotlib can will only draw contours for triangles e.g.).

The following suffices for Ugrid2d topologies in the _PlotMethods accessor:
```python

    def to_holoviews(self):
        import holoviews
        import pandas as pd

        grid = self.grid
        da = self.darray
        (_, _, triangles), index = grid.triangulation
        name = getattr(da, "name", "data")

        vertices = pd.DataFrame(
            data=grid.node_coordinates,
            columns=("x", "y"),
        )
        triangles = pd.DataFrame(
            data=triangles,
            columns=("v0", "v1", "v2"),
        )

        dim = get_ugrid_dim(grid, da)
        if dim == NODE:
            vertices[name] = da
        elif dim == FACE:
            triangles[name] = da.isel({grid.face_dimension: index})
        else:
            raise ValueError("only supports data on nodes or faces")

        return holoviews.TriMesh((triangles, vertices))
```

Then, to draw something:

```python
import holoviews as hv
from datashader import rasterize

hv.extension("matplotlib")
rasterize(uda.to_holoviews()).opts(colorbar=True)
```

What makes it not totally trivial:

* What to do for 1D topologies? Return a network I guess? https://datashader.org/user_guide/Networks.html
* The datashader/holoviews isn't entirely obvious to me. Should we add (more) methods to let it resemble the "standard" plotting arguments (i.e. like xarray/matplotlib?)
* Should we default to bokeh/matplotlib as the extension? Not setting it results in nothing being shown.

Secondly, there's a package called `hvplot`: https://hvplot.holoviz.org/
It adds an `hvplot` and `interactive` accessor to dataframes and xarray objects. This seems like it's more work, because it actually does data selection.

**Comment by rsignell** [comment]
Yes, @Huite `hvplot` was created to make it easier to plot data in Pandas dataframes or Xarray Dataarrays much easier:  `xarray` already had `.plot`, and now we have `.hvplot()` that works also!   

I use holoviz via hvplot exclusively with large model datasets for it's interactivity.   It's awsome. 

**But,** it doesn't currently work with unstructured grid xarray dataarrays.   

We have tried several times with Jim Bednar (@jbednar) to get funding to add support for UGRID compliant data, and were not successful.   But if we rely on xugrid data structures, it seems it would be quite straightforward, right?

BTW, regarding the design if implemented, I would enable rasterize by default when using TriMesh -- only smaller meshes will work well without it, and there really isn't much of a downside to enabling it (except the extra dependencies).

**Comment by rsignell** [comment]
I didn't really know about `xugrid` when we proposed adding the hvplot capability to `uxarray`.    I think it would be easier/better to add the functionality to `xugrid`. 

@ocefpaf took a look at the various libs that work with UGRID and concluded:

> with uxarray, IO is based on files, complex if-clauses to read exodus, shape file, scrip, and other formats, code based on multiple xarray reads of the same data, making it quite inefficient. Xugrid in the other hand seems to do the right thing. Outsources I/O to xarray, reads the metadata, creates the ugrid high level object.

I've been using `xugrid` lately and it's awesome!

**Comment by rsignell** [comment]
@jbednar and @Huite , I forked hvplot and asked Claude Code to create a new extension for `hvplot.xugrid` so that we could `hvplot.trimesh()` on a xugrid dataarray object.    

It worked!   Here's an example of it in action:
https://nbviewer.org/gist/rsignell/2ee54d379b15be4b8a101c5dec68bd6d

<img width="1255" height="819" alt="Image" src="https://github.com/user-attachments/assets/eb09f11b-9c4b-48ad-bd3e-273b5d02ecd5" />

The "enhanced" hvplot code is at:
https://github.com/OpenScienceComputing/hvplot/tree/xugrid

But I'm not sure what to do with it.  @jbednar should I submit this as a PR or leave it as a fork or...  ?

**Comment by rsignell** [comment]
The problem is, I don't really write code. I do scientific analysis and viz and that's about it!

**Comment by Huite** [comment]
From my perspective and at first glance, the code looks good -- though I guesss it needs some tests. It's a matter of glueing things together, and I would expect Claude to be perfectly capable in such cases. At first glance, this looks at least as good as my first attempt at submitting an xugrid PR to hvplot.
My primary worry would be the eager loading of data, but it looks like the [last commit](https://github.com/OpenScienceComputing/hvplot/commit/2b5746f9e3c60f5eb59251a4bd6c64b3e02ba7b7) tackles that issue -- @rsignell: I guess after some feedback from your side?

I'd be happy to see xugrid represented in hvplot, and I'll note a few things here (wonderful how a concrete proposal sharpens things!):

* I've styled xugrid as an ["xarray accessor"](https://docs.xarray.dev/en/stable/internals/extending-xarray.html); this is not yet reality. I currently wrap xarray in a new class instead given that xarray didn't have the capability, but it does now; see #35, and I'll be able to stuff the topology in a custom index. Perhaps something I, too, should delegate to Claude or Codex... In terms of hvplot it wouldn't mean much, just the `is_xugrid` check should defer to `.ugrid.is_ugrid` rather than a type check. If we move forward with this, I should probably just do the explicit indexes form.
* hvplot is more general than just xarray also supports pandas, etc., but the patching is basically the same as the xarray accessor registration.
* Anyway, this creates a bit of an "accessor collision". In my thinking for xugrid, the idea is that you can treat it as standard xarray, unless you want to do something with the topology in which case you insert `.ugrid`. In that case, the natural API form is `.ugrid.hvplot`, since calling it via `.hvplot` should treat it as an ordinary xarray structure without awareness of the unstructured topology.
* I could add my own `.hvplot` accessor to xugrid, but that feels dubious since I'd be "pirating" another package's namespace and I'd possibly have to import and rely on private classes and methods. hvplot on the other hand can simply rely on the public xugrid methods.
* Aternatively, hvplot could contains some logic for figuring whether it needs to create a spatial plot, and if so, use the topology (and otherwise treat it as ordinary xarray). This is probably better from a user perspective. (Zen of Python also states that "Flat is better than nested").

@rsignell: I'd suggest submitting it as a PR as @jbednar suggests. That'll give us an easy complete overview of the changes.

Before merging, I should finalize the explicit indexes so we don't start off with (minor) technical debt in an external (to me) package.

Secondly, there are still relevant additions for Holoviews/Datashader, such as more direct support for face data, possibly edge data (instead of only nodal) as well as alternative plotting methods (e.g. edge data will often represent velocities or flows). Such functionality is relevant to any unstructured topology package (such as uxarray), and is where I think funding should be spent primarily (if procured).

**Comment by rsignell** [comment]
@Huite , is this sufficient?  
https://github.com/pydata/xarray/pull/11286

**Comment by Huite** [comment]
I've seen it pop up in my notifications, thanks for making the effort!
And I see that there's been a new xarray release, great! I'll resume my testing tomorrow.

**Comment by rsignell** [comment]
@Huite , great!   I'm teaching a class next week in Italy using my fork so they can use hvplot.trimesh() with Xugrid, but I'll tell them that a better solution is in the works!

**Comment by rsignell** [comment]
BTW, I asked Claude about whether we should have `temp_surf.hvplot.trimesh()` or `temp_surf.ugrid.hvplot.trimesh()`.   What do you think we should do?

<img width="1873" height="708" alt="Image" src="https://github.com/user-attachments/assets/0c1557d5-6a02-4b62-8beb-49cd6f856ebe" />

**Comment by rsignell** [comment]
@Huite I'm also hoping https://github.com/pydata/xarray/pull/11286 will keep this conversation going.  Let me know if I can help!

**Comment by rsignell** [comment]
@Huite as you can see I decided to submit the PR to hvplot.  I did add tests which can be run with `pytest hvplot/tests/testxugrid.py`.

I'm giving a Pangeo Community Lightning Talk on xugrid and hvplot.xugrid tomorrow, so at least wanted to point to something on the hvplot repo!

**Comment by Huite** [comment]
Ah, good to know! As mentioned, I think this is all very worthwhile, so I'd be happy to (co-)develop and maintain -- I just have a number of deadlines at the moment...

Good luck with the talk!
